bluesky_kafka/mongo_normalized_consumer.py

```diff
@@ -0,0 +1,59 @@
+import cachetools
+import msgpack
+import os
+
+from bluesky_kafka import RemoteDispatcher
+from functools import partial
+from suitcase.mongo_normalized import Serializer
+
+bootstrap_servers = os.environ.get("KAFKA_BOOTSTRAP_SERVERS")
+mongo_uri = os.environ.get("BLUESKY_MONGO_URI")
+kafka_deserializer = partial(msgpack.loads, object_hook=mpn.decode)
+auto_offset_reset = "latest"  # "latest" should always work but has been failing on Linux, passing on OSX
+topics = ["^*.bluesky.documents"]
+
+
+class LazyMongoRouter():
+    """
+    Route messages to the correct mongo database, determined by the topic name.
+    TODO add caching mechanism to close old serializers.
+    """
+    def __init__(self, mongo_uri):
+        self._mongo_uri = mongo_uri
+        self._serializers = {}
+
+    @property
+    def serializers(self):
+        return self._serializers
+
+    def _get_serializer(self, topic):
+        database_name = topic.replace('.', '-')
+        serializer = Serializer(self._mongo_uri + database_name,
+                                self._mongo_uri + datamase_name)
+        self._serializers[topic] = serializer
+        return serializer
+
+    def __call__(self, topic, name, doc):
```

Comment:
Is `cb(topic, name, doc)` the signature that `kafka_dispatcher.subscribe` expects? I thought it just expected `cb(name, doc)`.

Reply:
I changed the bluesky_kafka.RemoteDispatcher so that it passes the topic name through.  At least I think the changes that I made should do that.

Reply:
I ask because making that change would split the callback ecosystem: we would have many callbacks that expect `(name, doc)` and now some that expect `(topic, name, doc)`. That gives me pause, makes me want to consider alternatives.

RemoteDispatcher already filters on topics (in its `__init__`). For our current purposes, is it sufficient to do the filtering at that level, and then attaching Serializers directly? What's the advantage of lazy and/or dynamic creation of mongo URIs?

Reply:
The advantage of lazy creation is that a single consumer may be subscribed to more than one topic. Then if there a consumer re balance the topics that all consumers are subscribed to may change. So I didn't want each consumer to have an instance of each serializer. Also if a new topic is added I didn't want to have to change the code of the consumers by adding a new serializer.

Reply:
What if we didn't use the callback system?

Reply:
We could make a separate consumer group for each topic.  I think this is more fragile.  If one consumer in the group fails then only consumers for that beamline consumer group could replace it.  But if all of the beamlines are in the same larger consumer group, there are many more consumers that are available to replace it.   

We would have to manage the status of each consumer group independently. For each beamline/topic we would have to check on it to make sure there are consumers running on each machine.  We may have one consumer group for each task (like mongo sync) for each beamline.   Which would could be a lot consumer groups.     

Also every time that we add a new beamline/topic we have to create a new custom comsumer group. This may mean managing a config file per beamline, per task. But if there is one consumer group for all beamlines, for a single task. Then there is no config file, each consumer is capable of supporting any beamline.

Reply:
> If there a consumer re balance the topics that all consumers are subscribed to may change.

Can you explain that a little more? I thought the set of topics would be a thing that would rarely change, once we decided on them.

Reply:
When there are multiple consumer groups subscribed to a topic there will be more than one consumer per partition. For example if there are two consumer groups then each partition will have two consumers. Each consumer group keeps a separate position in the topic, and documents aren't removed from Kafka until each consumer group has processed the documents.

Reply:
All of that sounds reasonable, as far as I can see. I focused on whether we need to dynamically accept topics we haven't seen before and silently create new MongoDB databases for them (which what Serializers do, if the database doesn't already exist).

Once we're past the initial step of adding all the current beamlines, adding a new beamline will be a rare thing that we don't need to optimize for. I would argue that adding a new beamline should _intentionally_ be something that is hard to do by accident. The current approach would lead a typo in a Publisher to silently create a MongoDB database.

Reply:
@jklynch  and I configured the Kafka cluster not to create new topics automatically. So we need to explicitly use a command line tool to create the topic before it is possible to write messages to it.

Reply:
I see. So if a Publisher tries to publish to some non-blessed topic name, Kafka will reject that message?

Reply:
Right

Reply:
So should we not use the callback registry? And merge the RemoteDispatcher and LazyMongoRouter into a new class?

Reply:
Yes, this does sound like a job for a topic-aware RemoteDispatcher subclass, or else a class that has-a RemoteDispatcher.

Reply:
If we don't use the callback registry why do we want something that inherits from Dispatcher? What other features of dispatcher are useful to us?

Reply:
Not having thought about this in detail, code related to connection, event loop start/stop, and some parts of the callback handling seem like they will either be reused or rewritten. The proposal here is a callback registry where the callbacks are dynamically added, in roughly a `defaultdict`-ish sense. This could be done as a fully separate monolith or it could be done by wrapping `RemoteDispatcher` via inheritance or composition and adding that dynamic aspect. Either approach seems defensible at this point.